adapters/adform/adform.go

```diff
@@ -369,16 +369,44 @@ func openRtbToAdformRequest(request *openrtb.BidRequest) (*adformRequest, []erro
 
 	return &adformRequest{
 		adUnits:       adUnits,
-		ip:            request.Device.IP,
-		advertisingId: request.Device.IFA,
-		userAgent:     request.Device.UA,
+		ip:            getIPSafely(request.Device),
+		advertisingId: getIFASafely(request.Device),
+		userAgent:     getUASafely(request.Device),
 		isSecure:      secure,
 		referer:       referer,
-		userId:        request.User.BuyerUID,
+		userId:        getBuyerUIDSafely(request.User),
 		tid:           tid,
 	}, errors
 }
 
+func getIPSafely(device *openrtb.Device) string {
```

Comment:
I wish there were a simpler way to do this generically besides a `reflect` hack like:

```go
func getStringFieldSafely(i interface{}, field string) string {
    if i == nil {
        return ""
    }

    v := reflect.ValueOf(i)
    f := reflect.Indirect(v).FieldByName(field)
    return f.String()
}

...

ip:            getStringFieldSafely(request.Device, "IP"),
advertisingId: getStringFieldSafely(request.Device, "IFA"),
userAgent:     getStringFieldSafely(request.Device, "UA"),
userId:        getStringFieldSafely(request.User, "BuyerID"),
```

Reply:
you and me both :(... not sure if I'm just bad with Go and there are better ways to do things, but this language is painful sometimes.